### PR TITLE
[core][experimental] Restore DAG exceptions and fix some memory leaks

### LINKED ISSUE #45313: [core][experimental] Accelerated DAG does not handle application-level IOErrors correctly
### What happened + What you expected to happen

Accelerated DAG uses IOError to indicate a system error when a backing channel closes. This would not work properly if the application itself throws an IOError. We should replace IOError with a system-level error, e.g., ChannelIOError.

### Versions / Dependencies

3.0dev

### Reproduction script

--

### Issue Severity

None

### LINKED ISSUE #46132: [ADAG] Better handling for RayDAGTaskError
### Description

When ADAG execution encounters an exception, it throws `RayDAGTaskError`. Currently the creation and handling of `RayDAGTaskError` deviates from classic Ray in several ways:
* `RayDAGTaskError` does not inherit from `RayError` or `RayTaskError`
* `RayDAGTaskError` is not thrown when calling `ray.get(compiled_dag_ref)` but instead returned as return value, and user needs to check the type of the result to decide how to process.

The challenge we saw earlier when wrapping ADAG execution exception with `RayTaskError` was that somehow throwing an exception at `ray.get(compiled_dag_ref)` would prevent `ReadRelease()` from being called for the deserialized object destructor or cause asyncio DAG reader to hang. Specially `test_asyncio_exceptions` would fail.

We should consider better ways of handling `RayDAGTaskError`.

### Use case

_No response_

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Fixes a few different bugs:
- Restore behavior where exceptions are raised during `ray.get` (see #46132)
- Fix memory leak in asyncio backend - this would cause future `get`s to hang if results were zero-copy deserialized
- Rename AwaitableDAGOutput to CompiledDAGFuture and use future-like syntax to `await` the result
- Copies the shared memory buffer if it's an exception object to avoid zero-copy deserialization
- Use a custom ChannelError instead of IOError upon channel errors

## Related issue number

Closes #46132.
Closes #45313.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/_private/worker.py [hunk 1]
```diff
@@ -814,6 +814,7 @@ def get_objects(
         self,
         object_refs: list,
         timeout: Optional[float] = None,
+        return_exceptions: bool = False,
```
Comment (ruisearch42) [inline python/ray/_private/worker.py:817]: Add doc

File: python/ray/dag/compiled_dag_node.py [hunk 2]
```diff
@@ -126,6 +128,19 @@ def _prep_task(self, task: "ExecutableTask") -> None:
...
+        task_exception=True,
+    )
+    wrapped = RayTaskError(
+        function_name="do_exec_tasks",
+        traceback_str=backtrace,
+        cause=exc,
+    )
+    return wrapped
+
+
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:143]: Change to `RayChannelError` on L162

File: python/ray/dag/compiled_dag_node.py [hunk 3]
```diff
@@ -163,26 +178,11 @@ def _exec_task(self, task: "ExecutableTask", idx: int) -> bool:
         output_val = method(*resolved_inputs)
         output_writer.write(output_val)
     except Exception as exc:
-        # TODO(rui): consider different ways of passing down the exception,
-        # e.g., wrapping with RayTaskError.
-        output_writer.write(RayDAGTaskError(exc))
+        output_writer.write(_wrap_exception(exc))
```
Comment (ruisearch42) [inline python/ray/dag/compiled_dag_node.py:181]: Catch `RayChannelError` for writing to channel

**Review by ruisearch42 — COMMENTED** [review]
This is great! Thanks for the change!

Some minor comments only.

File: python/ray/dag/compiled_dag_node.py (hunk 3 above)
Comment (stephanie-wang) [reply python/ray/dag/compiled_dag_node.py:181]: Ah good calls, thanks!

File: python/ray/dag/tests/experimental/test_accelerated_dag.py [hunk 4]
```diff
@@ -574,7 +652,7 @@ def test_dag_fault_tolerance(ray_start_regular_shared):
     for i in range(99):
         ref = compiled_dag.execute(1)
         res = ray.get(ref)
-        if isinstance(res[0], RayDAGTaskError):
+        if isinstance(res[0], RayTaskError):
```
Comment (ruisearch42) [inline python/ray/dag/tests/experimental/test_accelerated_dag.py:655]: We don't expect ray.get() to throw?

File: src/ray/core_worker/experimental_mutable_object_manager.cc [hunk 5]
```diff
@@ -340,6 +372,7 @@ Status MutableObjectManager::ReadRelease(const ObjectID &object_id)
 }
 
 Status MutableObjectManager::SetError(const ObjectID &object_id) {
+  RAY_LOG(DEBUG) << "Set error " << object_id;
```
Comment (ruisearch42) [inline src/ray/core_worker/experimental_mutable_object_manager.cc:375]: Nit: reader might think something went wrong, how about "SetError on " << object_id

**Review by ruisearch42 — APPROVED** [review]
Nice!

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 4 above)
Comment (stephanie-wang) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:657]: Er sorry, what do you mean? We want ray.get() to throw an error?

File: src/ray/core_worker/experimental_mutable_object_manager.cc (hunk 5 above)
Comment (stephanie-wang) [reply src/ray/core_worker/experimental_mutable_object_manager.cc:393]: Ah will change, but FYI this is DEBUG level so users shouldn't see it.

File: python/ray/dag/tests/experimental/test_accelerated_dag.py (hunk 4 above)
Comment (ruisearch42) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:657]: In other cases error is thrown: https://github.com/ray-project/ray/pull/46154/files#diff-95fd0ac9e0e48063233cc0b007e51fcad639e8d5cf5391c446deace83506e1ffR382
But why not here? I think I may be missing something.

Comment (stephanie-wang) [reply python/ray/dag/tests/experimental/test_accelerated_dag.py:657]: Oh I see, good catch! Let me look into that.

**Pull request merged**